File: src/main/java/xyz/emirdev/echogen/Echogen.java
package xyz.emirdev.echogen;

import lombok.Getter;
import org.bukkit.Bukkit;
import org.bukkit.event.Listener;
import org.bukkit.plugin.java.JavaPlugin;

import io.papermc.paper.command.brigadier.Commands;
import io.papermc.paper.plugin.lifecycle.event.types.LifecycleEvents;
import net.luckperms.api.LuckPerms;
import net.luckperms.api.LuckPermsProvider;
import xyz.emirdev.echogen.database.PrefixDatabase;
import xyz.emirdev.echogen.managers.FilterManager;
import xyz.emirdev.echogen.managers.PrefixManager;
import xyz.emirdev.echogen.managers.ScoreboardManager;
import xyz.emirdev.echogen.managers.VanishManager;
import xyz.emirdev.echogen.task.BossBarTask;
import xyz.emirdev.echogen.utils.ClassUtils;
import xyz.emirdev.echogen.utils.MiniMessageUtils;

public class Echogen extends JavaPlugin {
    private static Echogen instance;
    @Getter
    private LuckPerms luckPerms;
    private PluginConfig config;
    @Getter
    private ScoreboardManager scoreboardManager;
    @Getter
    private VanishManager vanishManager;
    @Getter
    private FilterManager filterManager;
    @Getter
    private PrefixManager prefixManager;
    @Getter
    private PrefixDatabase prefixDatabase;
    @Getter
    private MiniMessageUtils miniMessageUtils;
    @Getter
    private boolean isPAPIEnabled;
    @Getter
    private boolean isSkriptEnabled;

    public static Echogen get() {
        return instance;
    }

    public PluginConfig getPluginConfig() {
        return config;
    }

    public void reloadConfig() {
        this.config.load();
        scoreboardManager.toggle(config.getRootNode().node("scoreboard", "enabled").getBoolean());
        filterManager.load();
        prefixManager.load();
    }

    @Override
    public void onEnable() {
        instance = this;
        luckPerms = LuckPermsProvider.get();
        isPAPIEnabled = Bukkit.getPluginManager().isPluginEnabled("PlaceholderAPI");
        isSkriptEnabled = Bukkit.getPluginManager().isPluginEnabled("Skript");

        config = new PluginConfig(this);
        config.load();

        scoreboardManager = new ScoreboardManager();
        Bukkit.getPluginManager().registerEvents(scoreboardManager, this);

        vanishManager = new VanishManager();
        Bukkit.getPluginManager().registerEvents(vanishManager, this);

        filterManager = new FilterManager();

        prefixManager = new PrefixManager();
        prefixDatabase = new PrefixDatabase();

        miniMessageUtils = new MiniMessageUtils();
        miniMessageUtils.run();

        BossBarTask.startAll();

        if (isPAPIEnabled)
            new PAPIExpansion().register();

        this.getLifecycleManager().registerEventHandler(LifecycleEvents.COMMANDS, commands -> {
            Commands registrar = commands.registrar();
            ClassUtils.findClasses(
                    "xyz.emirdev.echogen.commands",
                    clazz -> clazz.extendsSuperclass(PluginCommand.class),
                    clazz -> {
                        PluginCommand command;
                        try {
                            command = clazz.loadClass().asSubclass(PluginCommand.class).getConstructor().newInstance();
                        } catch (Exception e) {
                            throw new RuntimeException(e);
                        }

                        registrar.register(command.getCommand(), command.getAliases());
                    }
            );
        });

        ClassUtils.findClasses(
                "xyz.emirdev.echogen.events",
                clazz -> clazz.implementsInterface(Listener.class),
                clazz -> {
                    try {
                        Bukkit.getPluginManager().registerEvents((Listener) clazz.loadClass().getConstructor().newInstance(), this);
                    } catch (Exception e) {
                        throw new RuntimeException(e);
                    }
                }
        );
    }

    @Override
    public void onDisable() {
        instance = null;
    }
}

File: src/main/kotlin/cat/emir/echogen/Echogen.kt
package cat.emir.echogen

import cat.emir.echogen.database.PrefixDatabase
import cat.emir.echogen.managers.FilterManager
import cat.emir.echogen.managers.PrefixManager
import cat.emir.echogen.managers.ScoreboardManager
import cat.emir.echogen.managers.VanishManager
import cat.emir.echogen.task.BossBarTask
import cat.emir.echogen.utils.LuckPermsUtils
import cat.emir.echogen.utils.MiniMessageUtils
import cat.emir.echogen.utils.TimeUtils
import cat.emir.echolib.EchoPlugin
import cat.emir.echolib.PluginConfig
import cat.emir.echolib.command.CommandLib
import cat.emir.echolib.event.EventLoader
import com.mojang.brigadier.arguments.StringArgumentType
import com.mojang.brigadier.context.CommandContext
import io.papermc.paper.command.brigadier.CommandSourceStack
import net.luckperms.api.LuckPerms
import net.luckperms.api.LuckPermsProvider
import java.time.Duration

fun CommandContext<CommandSourceStack>.getDuration(argument: String): Duration? {
    val durationString = StringArgumentType.getString(this, argument)
    val duration = TimeUtils.convertStringToDuration(durationString)

    if (duration == null) {
        this.source.sender.sendRichMessage("<red>Invalid duration provided</red>")
        return null
    }

    return duration
}

class Echogen : EchoPlugin() {
    companion object {
        lateinit var instance: Echogen
            private set
    }
    
    lateinit var luckPerms: LuckPerms
        private set

    val config = PluginConfig(this, "config.yml")
    val scoreboardManager = ScoreboardManager(this)
    val vanishManager = VanishManager(this)
    val filterManager = FilterManager(this)
    val prefixManager = PrefixManager(this)
    val prefixDatabase = PrefixDatabase(this, "database")
    val miniMessageUtils = MiniMessageUtils(this)
    val luckPermsUtils = LuckPermsUtils(this)
    var isPAPIEnabled = false
    var isSkriptEnabled = false
    var isOpenInvEnabled = false

    override fun reloadConfig() {
        this.config.load()
        scoreboardManager.toggle(config.rootNode.node("scoreboard", "enabled").boolean)
        filterManager.load()
        prefixManager.load()
    }

    override fun onEnable() {
        instance = this
        luckPerms = LuckPermsProvider.get()
        isPAPIEnabled = server.pluginManager.isPluginEnabled("PlaceholderAPI")
        isSkriptEnabled = server.pluginManager.isPluginEnabled("Skript")
        isOpenInvEnabled = server.pluginManager.isPluginEnabled("OpenInv")

        config.load()

        scoreboardManager.load()
        server.pluginManager.registerEvents(scoreboardManager, this)

        vanishManager.load()
        server.pluginManager.registerEvents(vanishManager, this)

        filterManager.load()

        prefixManager.load()
        prefixDatabase.load()

        miniMessageUtils.load()

        BossBarTask.startAll(this)

        if (isPAPIEnabled)
            PAPIExpansion(this).register()

        CommandLib.registerCommands(this, "cat.emir.echogen.commands")
        EventLoader.registerEvents(this, "cat.emir.echogen.events")
    }
}
